third_party/move/move-compiler-v2/src/pipeline/uninitialized_use_checker.rs

```diff
@@ -0,0 +1,381 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+//! Implements a checker which verifies that all locals are initialized before any use.
+//! This checker does not require any other analysis to be run before.
+//! As a side effect of running this checker, function targets are annotated with
+//! the initialized state (yes, no, maybe) of all locals at each program point.
+//!
+//! There are two parts to this checker:
+//! * `InitializedStateAnalysis` which computes the initialized state of all locals at each
+//!   program point via a forward dataflow analysis.
+//! * `UninitializedUseChecker` which checks that all locals are initialized before use.
+
+use im::Vector;
+use move_binary_format::file_format::CodeOffset;
+use move_model::{ast::TempIndex, model::FunctionEnv};
+use move_stackless_bytecode::{
+    dataflow_analysis::{DataflowAnalysis, TransferFunctions},
+    dataflow_domains::{AbstractDomain, JoinResult},
+    function_target::{FunctionData, FunctionTarget},
+    function_target_pipeline::{FunctionTargetProcessor, FunctionTargetsHolder},
+    stackless_bytecode::Bytecode,
+    stackless_control_flow_graph::StacklessControlFlowGraph,
+};
+use std::collections::BTreeMap;
+
+/// State of initialization of a local at a given program point.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum Initialized {
```

Comment:
Perhaps we can we implement some common constructions on lattice. In this case, a general construction could be given a set T, attach a top element and a bottome element to it, such that for all t in T, bot <= t, t <= top, and nothing else. Another common one might be lattice products, which perhaps already appeared in the code base, where an `AbstractDomain` has other `AbstractDomain` fields and the join is fieldwise join.

Reply:
https://github.com/aptos-labs/aptos-core/issues/11632

Reply:
Agree, see comment reply on a similar comment below.